eclipse-theia/theia
# Refactor preference UI + add 'Commonly Used' section

#### What it does

Fixes: #9138 by adding a 'Commonly Used' section.
Fixes: #9438

 ##### Easily visible changes:
 - Cleans up left-hand-tree:
<table>
<tbody>
<tr>
<th>Before</th><th>After</th>
</tr>
<tr>
<td><img src="https://user-images.githubusercontent.com/62660806/117169878-1fc70f80-ad8f-11eb-8cf8-3d6a038e0333.png"/></td>
<td><img src="https://user-images.githubusercontent.com/62660806/117169912-26558700-ad8f-11eb-8915-39b512e83a56.png"/></td>
</tr>
</tbody>
</table>
- Adds 'Commonly Used' section:
![image](https://user-images.githubusercontent.com/62660806/117170346-8ea46880-ad8f-11eb-84c0-e5b65dadc4cd.png)
- Removes headlines and shows full ID path when filtering
<table>
<tbody>
<tr>
<th>Before</th><th>After</th>
</tr>
<tr>
<td><img src="https://user-images.githubusercontent.com/62660806/117170605-c90e0580-ad8f-11eb-83f2-22ecbe4ad291.png"/></td>
<td><img src="https://user-images.githubusercontent.com/62660806/117170649-d0cdaa00-ad8f-11eb-890c-cc57e0ae4564.png"/></td>
</tr>
</tbody>
</table>
 - Makes the 'Modified in...' messages clickable - clicking on one will set the search bar to the id of the preference and switch to the named scope.

##### Less visible changes
 - All React code in the `PreferencesEditorWidget` replaced with DOM methods.
 - Renderers now `@injectable()` and able to take advantage of and be instantiated by dependency injection.
 - Various changes to interfaces to make better use of Theia built-ins.
 - Addition of `PreferenceTreeLabelProvider` to allow for dynamic headline generation.
 - A few changes in other packages
   - Added a `DEFAULT_SCROLL_OPTIONS` constant to `widget.ts`. In all cases in which a widget uses scroll options in the framework, they use the same settings, but I haven't converted those to references to `DEFAULT_SCROLL_OPTIONS`.
   - Added a static `.deepEqual()` method to the `PreferenceProvider` class for use in `PreferenceProviders` and the node renderers in the preferences package.
   - Added a default type `JSONValue` to the generic on the `PreferenceInspection` type, since it always will be a `JSONValue`.

#### How to test

1. Use the preferences UI in its various use cases:
 - Set preferences
   - In particular, use string and number inputs, and ensure that there are no jerky interactions with incoming preference change events.
 - Change scopes - make sure the values update and the headlines correctly reflect other modified scopes.
 - Search for particular values - make sure the display is correct and reasonably fast
 - Search and change scopes - make sure that everything that should happen in both scenarios happens when both are active together.
 - Click on a 'Modified in...' message and see that both scope and searchbar are updated.

#### Review checklist


#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

Signed-off-by: Colin Grant <colin.grant@ericsson.com>
